x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportResetJobAction.java

```diff
@@ -0,0 +1,232 @@
+            indexNameExpressionResolver, ThreadPool.Names.SAME);
+        this.client = Objects.requireNonNull(client);
+        this.jobConfigProvider = Objects.requireNonNull(jobConfigProvider);
+        this.jobResultsProvider = Objects.requireNonNull(jobResultsProvider);
+        this.auditor = Objects.requireNonNull(auditor);
+    }
+
+    @Override
+    protected void masterOperation(Task task, ResetJobAction.Request request, ClusterState state,
+                                   ActionListener<AcknowledgedResponse> listener) throws Exception {
+        final TaskId taskId = new TaskId(clusterService.localNode().getId(), task.getId());
+
+        ActionListener<Job.Builder> jobListener = ActionListener.wrap(
+            jobBuilder -> {
+                Job job = jobBuilder.build();
+                PersistentTasksCustomMetadata tasks = state.getMetadata().custom(PersistentTasksCustomMetadata.TYPE);
+                JobState jobState = MlTasks.getJobState(job.getId(), tasks);
+                if (request.isForce() == false && jobState != JobState.CLOSED) {
+                    listener.onFailure(ExceptionsHelper.conflictStatusException(Messages.getMessage(Messages.REST_JOB_NOT_CLOSED_RESET)));
+                    return;
+                }
+                if (job.getBlockReason() != null && job.getBlockReason() != BlockReason.RESET) {
+                    listener.onFailure(ExceptionsHelper.conflictStatusException(
+                        "cannot reset job while it is blocked with [" + job.getBlockReason() + "]"));
+                    return;
+                }
+
+                ParentTaskAssigningClient taskClient = new ParentTaskAssigningClient(client, taskId);
+                jobConfigProvider.updateJobBlockReason(job.getId(), BlockReason.RESET, ActionListener.wrap(
+                    aBoolean -> resetJob(taskClient, (CancellableTask) task, request, listener),
+                    listener::onFailure
+                ));
+            },
+            listener::onFailure
+        );
+
+        ActionListener<TaskInfo> existingTaskListener = ActionListener.wrap(
+            existingTask -> {
+                if (existingTask == null) {
+                    jobConfigProvider.getJob(request.getJobId(), jobListener);
+                } else {
+                    waitExistingResetTaskToComplete(existingTask, request, listener);
+                }
+            },
+            listener::onFailure
+        );
+
+        findExistingResetTask(taskId, request.getJobId(), existingTaskListener);
+    }
+
+    private void findExistingResetTask(TaskId currentTaskId, String jobId, ActionListener<TaskInfo> listener) {
+        ListTasksRequest listTasksRequest = new ListTasksRequest();
+        listTasksRequest.setActions(ResetJobAction.NAME);
+        listTasksRequest.setDescriptions(MlTasks.JOB_TASK_ID_PREFIX + jobId);
+        listTasksRequest.setDetailed(true);
+        listTasksRequest.setNodes(currentTaskId.getNodeId());
```

Comment:
🔧

What if the master node has changed? Is it possible for the master node to change without the previous master node stopping?

This may only occur for a smaller period of time when a cluster first starts, or if there are network connection issues...So it may not be a concern.

Reply:
I think when it comes to best effort this is all we can do. The worse case scenario is we launch a second reset which isn't catastrophic.